File: src/lib/detector-envelope.ts
/**
 * detector-envelope.ts — OPS-MONITORING-SIGNAL-CONTRACT-W1 CH2, PRODUCER side.
 *
 * A MONITORING DETECTOR IS A VERIFICATION GATE THAT NOBODY CALLED A GATE. It reads state, emits a
 * verdict, and a human acts on it — so every law written for `scripts/check-*` applies to it, and
 * the one that mattered is that "measured and short" must never share an output with "never
 * finished measuring."
 *
 * `detectCapacityShortfall` published a structural capacity verdict from a run SIGTERM'd at 46.6
 * of 210 minutes because it had no way to say INDETERMINATE. This module makes that emission
 * IMPOSSIBLE rather than discouraged: `buildEnvelope` FORCES the verdict to INDETERMINATE when the
 * run outcome is not one the schema calls conclusive, so a producer cannot state a conclusion its
 * own run did not reach — no matter what the caller passes.
 *
 * THE SCHEMA IS THE SoT AND IT IS DATA, NOT CODE. `ops/monitoring/detector-envelope.schema.json`
 * is read here AND by `ops/monitoring/detector_envelope.py`. A shared TypeScript helper would
 * leave the Python consumer unconstrained, which is precisely where D2 and D3 lived.
 *
 * ── WHY THERE IS AN EMBEDDED MIRROR (OPS-DETECTOR-ENVELOPE-RUNTIME-W1) ──
 *
 * That SoT file is STRUCTURALLY ABSENT from the runtime image. The Dockerfile COPYs no `ops/`
 * path, so `/app/ops` does not exist, and every in-container `buildEnvelope` call threw ENOENT
 * instead of emitting a verdict. Measured 2026-08-26: 18 throws in /var/log/carry-labeler.log,
 * each one killing the nightly labeler AFTER it had written its rows, and the consumer
 * (`directional-label-freshness.py`) reporting `CAPACITY_SIGNAL INDETERMINATE` every morning
 * since 2026-08-23 because the producer had gone silent.
 *
 * SHIPPING THE FILE INTO THE IMAGE WAS PROBED AND REJECTED, and the reason is a measured platform
 * limit rather than a preference. A COPYed path must leave `deploy.yml`'s `paths-ignore` or a
 * schema edit ships a stale image — and GitHub Actions cannot express "ignore `ops/monitoring/**`
 * except this one file": `!` negation works only in `paths`, and `paths` and `paths-ignore`
 * cannot be combined for one event. The reachable alternatives were dropping `ops/monitoring/**`
 * wholesale (every canary edit then rebuilds and restarts prod — the exact thing that entry
 * exists to prevent) or inverting the trigger to a `paths` allow-list, whose failure mode is
 * worse than this defect: a future wave adding a runtime path would silently never deploy.
 *
 * SO THE MIRROR IS NOT A SECOND SoT AND IT IS NOT A PERMISSIVE DEFAULT. It is a projection of the
 * same file, pinned FIELD-FOR-FIELD by `tests/unit/detector-envelope.test.ts`, which also asserts
 * the JSON declares no contract key the mirror lacks. A schema edit that is not mirrored fails the
 * pre-push gate and CI — strictly stronger than a rebuild trigger, which only fires if a deploy
 * happens to run. This is the same shape the Python consumer already uses: mirror the rule, pin
 * the pair with one shared-corpus test.
 *
 * `ops/monitoring/detector_envelope.py::load_schema` deliberately does NOT do this, and the
 * asymmetry is the point. It is the CONSUMER — a validator that invented its own contract would
 * accept envelopes nobody could have emitted, which is the dark-guard class. This is the
 * PRODUCER: it emits `schema_version`, so a mirror that ever did drift is rejected downstream by
 * the consumer's own version check rather than laundered into a pass.
 */
import { readFileSync } from 'node:fs';
import * as path from 'node:path';

export type Verdict = 'PASS' | 'FAIL' | 'INDETERMINATE';

export interface DetectorEnvelope {
  schema_version: number;
  detector: string;
  verdict: Verdict;
  run_id: string;
  run_started_at: string;
  run_outcome: string;
  produced_at: string;
  observation_window: { from: string; to: string };
  evidence: Record<string, unknown>;
}

export interface EnvelopeSchema {
  schema_version: number;
  required_fields: string[];
  verdict_values: Verdict[];
  verdict_default: Verdict;
  run_outcome_values: string[];
  run_outcome_conclusive: string[];
  max_age_seconds: number;
  evidence_rules: { must_be_object: boolean; min_keys: number; max_prose_words: number };
  observation_window_fields: string[];
}

/**
 * Resolved from THIS file's location. Correct in the repo and on any host that has the SoT beside
 * it; in the runtime image it resolves to `/app/ops/monitoring/...`, which does not exist — see
 * the module header for why that is answered by the mirror below and not by a Dockerfile COPY.
 */
export function schemaPath(): string {
  return path.resolve(__dirname, '../../ops/monitoring/detector-envelope.schema.json');
}

/**
 * The contract, mirrored. EVERY key of `EnvelopeSchema` and nothing else — the `_comment`,
 * `_generator` and `_*_doc` keys of the JSON are documentation and are deliberately not carried.
 * `tests/unit/detector-envelope.test.ts` asserts this equals the SoT field for field AND that the
 * SoT declares no non-underscore key missing here, so a contract change cannot be pushed
 * half-applied.
 */
export const EMBEDDED_SCHEMA: EnvelopeSchema = Object.freeze({
  schema_version: 1,
  required_fields: Object.freeze([
    'schema_version',
    'detector',
    'verdict',
    'run_id',
    'run_started_at',
    'run_outcome',
    'produced_at',
    'observation_window',
    'evidence',
  ]) as unknown as string[],
  verdict_values: Object.freeze(['PASS', 'FAIL', 'INDETERMINATE']) as unknown as Verdict[],
  verdict_default: 'INDETERMINATE',
  run_outcome_values: Object.freeze([
    'complete',
    'venue-budget',
    'global-budget',
    'venue-error',
    'venue-circuit-break',
    'stopped',
    'unknown',
  ]) as unknown as string[],
  run_outcome_conclusive: Object.freeze(['complete', 'venue-budget', 'global-budget']) as unknown as string[],
  max_age_seconds: 21600,
  evidence_rules: Object.freeze({ must_be_object: true, min_keys: 1, max_prose_words: 7 }),
  observation_window_fields: Object.freeze(['from', 'to']) as unknown as string[],
}) as EnvelopeSchema;

let cached: EnvelopeSchema | null = null;
let announcedMirror = false;

/**
 * THE SUBSTITUTION IS NEVER SILENT, AND IT IS NEVER A WIDENING.
 *
 *   - default path, file present     → the file wins. It is the SoT.
 *   - default path, file ABSENT      → the pinned mirror, plus ONE stderr line per process.
 *   - default path, file UNREADABLE  → THROWS. Input we were handed and could not parse is the
 *                                      indeterminate case; a corrupt SoT must never be papered
 *                                      over by a copy that happens to still parse.
 *   - EXPLICIT path, missing         → THROWS. A caller naming a file is asserting something about
 *                                      THAT file, and answering with different bytes would make
 *                                      the assertion unfalsifiable.
 */
export function loadSchema(p: string = schemaPath()): EnvelopeSchema {
  const isDefault = p === schemaPath();
  if (cached && isDefault) return cached;

  let raw: string;
  try {
    raw = readFileSync(p, 'utf8');
  } catch (err) {
    const code = (err as NodeJS.ErrnoException)?.code;
    if (!isDefault || (code !== 'ENOENT' && code !== 'ENOTDIR')) throw err;
    if (!announcedMirror) {
      announcedMirror = true;
      console.error(
        `[detector-envelope] schema SoT absent at ${p}; using the embedded mirror `
        + '(pinned field-for-field by tests/unit/detector-envelope.test.ts). '
        + 'Expected inside the runtime image, which COPYs no ops/ path.',
      );
    }
    if (isDefault) cached = EMBEDDED_SCHEMA;
    return EMBEDDED_SCHEMA;
  }

  const s = JSON.parse(raw) as EnvelopeSchema;
  if (isDefault) cached = s;
  return s;
}
/** Test seam — the module-level cache would otherwise pin the first schema a suite loaded. */
export function _resetSchemaCacheForTest(): void { cached = null; announcedMirror = false; }

export interface BuildInput {
  detector: string;
  verdict?: Verdict;
  runId: string;
  runStartedAt: string;
  runOutcome: string;
  producedAt: string;
  observationWindow: { from: string; to: string };
  evidence: Record<string, unknown>;
}

/**
 * Build a conforming envelope. THE FORCING RULE IS THE WHOLE POINT:
 *
 *   - an omitted verdict becomes INDETERMINATE (the schema's declared default), never PASS;
 *   - a run whose `run_outcome` is NOT in `run_outcome_conclusive` has its verdict DOWNGRADED to
 *     INDETERMINATE even when the caller passed PASS or FAIL.
 *
 * So the 2026-08-22 run — `outcome=stopped`, 46.6 of 210 minutes — cannot produce a FAIL saying
 * the budget is structurally short. The truncation stays fully visible in `evidence` and
 * `run_outcome`; only the conclusion the run did not earn is removed.
 */
export function buildEnvelope(input: BuildInput, schema: EnvelopeSchema = loadSchema()): DetectorEnvelope {
  const asked: Verdict = input.verdict ?? schema.verdict_default;
  const conclusive = schema.run_outcome_conclusive.includes(input.runOutcome);
  const verdict: Verdict = conclusive ? asked : 'INDETERMINATE';
  return {
    schema_version: schema.schema_version,
    detector: input.detector,
    verdict,
    run_id: input.runId,
    run_started_at: input.runStartedAt,
    run_outcome: input.runOutcome,
    produced_at: input.producedAt,
    observation_window: input.observationWindow,
    evidence: input.evidence,
  };
}

/** Every reason an envelope is non-conforming. Empty array = conforming. */
export function validateEnvelope(env: unknown, schema: EnvelopeSchema = loadSchema()): string[] {
  const errs: string[] = [];
  if (env === null || typeof env !== 'object' || Array.isArray(env)) return ['envelope is not an object'];
  const e = env as Record<string, unknown>;
  for (const f of schema.required_fields) if (!(f in e)) errs.push(`missing required field '${f}'`);
  if ('schema_version' in e && e.schema_version !== schema.schema_version) {
    errs.push(`schema_version ${String(e.schema_version)} != ${schema.schema_version}`);
  }
  if ('verdict' in e && !schema.verdict_values.includes(e.verdict as Verdict)) {
    errs.push(`verdict '${String(e.verdict)}' is not one of ${schema.verdict_values.join('|')}`);
  }
  if ('run_outcome' in e && !schema.run_outcome_values.includes(String(e.run_outcome))) {
    errs.push(`run_outcome '${String(e.run_outcome)}' is not a declared outcome`);
  }
  // The forcing rule, asserted on the WIRE as well as at build time. A producer that hand-rolls an
  // envelope instead of calling buildEnvelope must not be able to smuggle the conclusion past us.
  if ('run_outcome' in e && 'verdict' in e
      && !schema.run_outcome_conclusive.includes(String(e.run_outcome))
      && e.verdict !== 'INDETERMINATE') {
    errs.push(`run_outcome '${String(e.run_outcome)}' is not conclusive, so verdict MUST be INDETERMINATE, not '${String(e.verdict)}'`);
  }
  const w = e.observation_window as Record<string, unknown> | undefined;
  if ('observation_window' in e) {
    if (w === null || typeof w !== 'object') errs.push('observation_window is not an object');
    else for (const f of schema.observation_window_fields) if (!(f in w)) errs.push(`observation_window missing '${f}'`);
  }
  if ('evidence' in e) {
    const ev = e.evidence as Record<string, unknown>;
    if (ev === null || typeof ev !== 'object' || Array.isArray(ev)) errs.push('evidence is not an object');
    else {
      if (Object.keys(ev).length < schema.evidence_rules.min_keys) {
        errs.push(`evidence has ${Object.keys(ev).length} key(s), fewer than the required ${schema.evidence_rules.min_keys}`);
      }
      // D2's defect, caught in the field that exists to prevent it: a sentence about mechanism is
      // not a measurement, and it keeps rendering long after the code beneath it changes.
      for (const [k, v] of Object.entries(ev)) {
        if (typeof v !== 'string') continue;
        const words = v.trim().split(/\s+/).filter(Boolean).length;
        if (words > schema.evidence_rules.max_prose_words) {
          errs.push(`evidence.${k} is ${words} words — prose about mechanism, not a measured value (max ${schema.evidence_rules.max_prose_words})`);
        }
      }
    }
  }
  return errs;
}

export function isConforming(env: unknown, schema: EnvelopeSchema = loadSchema()): boolean {
  return validateEnvelope(env, schema).length === 0;
}

File: ops/monitoring/detector_envelope.py
#!/usr/bin/env python3
"""detector_envelope.py — OPS-MONITORING-SIGNAL-CONTRACT-W1 CH2, CONSUMER side.

THE HALF THAT ACTUALLY STOPS D3. A producer-side contract alone would have changed nothing about
2026-08-22: `forward_capacity_signal` re-read a marker and forwarded it verbatim, so the refusal
has to live where the forwarding decision is made.

Reads the SAME SoT as the TypeScript producer — `detector-envelope.schema.json` — because the
producer is TS and the consumer is Python, and a shared code module could only ever constrain one
of them.

  decide(envelope, current_run_id, now) -> Decision(action, verdict, reason)

    FORWARD   the envelope is conforming, current, in-age, and its verdict is PASS or FAIL
    REFUSE    something is wrong with the SIGNAL, not with the system it describes

A REFUSAL IS VISIBLE. It resolves to INDETERMINATE and says why; it never resolves to silence, and
it never resolves to FAIL. An operator paged for "we could not measure" is reading a different
page from "we measured and it is bad", and collapsing the two is the defect this wave retires.

Exit: 0 = self-test passed · 1 = self-test failed · 3 = INDETERMINATE (could not run).
Callers gate on the TOKEN `DETECTOR_ENVELOPE_VERDICT=PASS|FAIL|INDETERMINATE`, never the code.
"""
from __future__ import annotations

import json
import os
import sys
from datetime import datetime, timezone
from pathlib import Path

SCHEMA_PATH = Path(os.environ.get(
    "DETECTOR_ENVELOPE_SCHEMA",
    str(Path(__file__).resolve().parent / "detector-envelope.schema.json"),
))

FORWARD = "FORWARD"
REFUSE = "REFUSE"


class Decision:
    __slots__ = ("action", "verdict", "reason")

    def __init__(self, action: str, verdict: str, reason: str) -> None:
        self.action, self.verdict, self.reason = action, verdict, reason

    def __repr__(self) -> str:  # pragma: no cover - diagnostics only
        return f"Decision({self.action}, {self.verdict}, {self.reason!r})"


def load_schema(path: Path = SCHEMA_PATH) -> dict:
    """An unreadable schema is INDETERMINATE, never a permissive default.

    A validator that falls back to "allow everything" when it cannot read its own contract is a
    dark guard: it exits 0 and is indistinguishable from a healthy one.
    """
    return json.loads(path.read_text())


def _parse_iso(ts: str):
    try:
        return datetime.fromisoformat(str(ts).replace("Z", "+00:00"))
    except Exception:
        return None


def validate(env, schema: dict) -> list:
    """Every reason this envelope is non-conforming. Empty list = conforming.

    Deliberately mirrors `src/lib/detector-envelope.ts::validateEnvelope` rule for rule; the pair
    is pinned by `tests/unit/detector-envelope.test.ts`, which runs BOTH sides over one shared
    fixture set so the two implementations cannot drift into disagreeing about the same bytes.
    """
    errs = []
    if not isinstance(env, dict):
        return ["envelope is not an object"]
    for f in schema["required_fields"]:
        if f not in env:
            errs.append("missing required field '%s'" % f)
    if "schema_version" in env and env["schema_version"] != schema["schema_version"]:
        errs.append("schema_version %s != %s" % (env["schema_version"], schema["schema_version"]))
    if "verdict" in env and env["verdict"] not in schema["verdict_values"]:
        errs.append("verdict '%s' is not one of %s" % (env["verdict"], "|".join(schema["verdict_values"])))
    if "run_outcome" in env and env["run_outcome"] not in schema["run_outcome_values"]:
        errs.append("run_outcome '%s' is not a declared outcome" % env["run_outcome"])
    # The forcing rule, asserted on the WIRE. A producer that hand-rolls an envelope must not be
    # able to smuggle a conclusion its own run never reached.
    if ("run_outcome" in env and "verdict" in env
            and env["run_outcome"] not in schema["run_outcome_conclusive"]
            and env["verdict"] != "INDETERMINATE"):
        errs.append("run_outcome '%s' is not conclusive, so verdict MUST be INDETERMINATE, not '%s'"
                    % (env["run_outcome"], env["verdict"]))
    if "observation_window" in env:
        w = env["observation_window"]
        if not isinstance(w, dict):
            errs.append("observation_window is not an object")
        else:
            for f in schema["observation_window_fields"]:
                if f not in w:
                    errs.append("observation_window missing '%s'" % f)
    if "evidence" in env:
        ev = env["evidence"]
        rules = schema["evidence_rules"]
        if not isinstance(ev, dict):
            errs.append("evidence is not an object")
        else:
            if len(ev) < rules["min_keys"]:
                errs.append("evidence has %d key(s), fewer than the required %d" % (len(ev), rules["min_keys"]))
            for k, v in ev.items():
                if not isinstance(v, str):
                    continue
                words = len([w for w in v.strip().split() if w])
                if words > rules["max_prose_words"]:
                    errs.append("evidence.%s is %d words — prose about mechanism, not a measured "
                                "value (max %d)" % (k, words, rules["max_prose_words"]))
    return errs


def decide(env, current_run_id, now=None, schema=None) -> Decision:
    """Should this envelope be forwarded to the operator?

    Order matters and is not arbitrary — CONFORMANCE first, then IDENTITY, then AGE, then VERDICT.
    Each check is strictly cheaper and strictly more certain than the next: an envelope we cannot
    parse tells us nothing about whether it is current, and one from the wrong run tells us
    nothing about whether it is fresh. Reporting the LAST failing reason instead of the first
    would name a symptom over a cause.
    """
    schema = schema if schema is not None else load_schema()
    now = now or datetime.now(timezone.utc)

    errs = validate(env, schema)
    if errs:
        return Decision(REFUSE, "INDETERMINATE", "envelope is non-conforming: %s" % "; ".join(errs[:3]))

    # IDENTITY — is this THIS run's signal? `current_run_id` of None means the consumer could not
    # determine the current run, which is itself a non-answer and must not be read as a match.
    if current_run_id is None:
        return Decision(REFUSE, "INDETERMINATE",
                        "consumer could not determine the current run id — cannot confirm this "
                        "signal describes the run it is about to page for")
    if env["run_id"] != current_run_id:
        return Decision(REFUSE, "INDETERMINATE",
                        "run_id %s is not the current run %s — this signal describes an older run"
                        % (env["run_id"], current_run_id))

    # AGE. NOTE: this is NOT the control that would have stopped the 2026-08-22 page — that marker
    # WAS from the most recent run. See `_max_age_doc` in the schema.
    produced = _parse_iso(env["produced_at"])
    if produced is None:
        return Decision(REFUSE, "INDETERMINATE", "produced_at %r is not parseable ISO-8601" % env["produced_at"])
    age = (now - produced).total_seconds()
    if age > schema["max_age_seconds"]:
        return Decision(REFUSE, "INDETERMINATE",
                        "produced_at is %.0fs old, past the %ds bound" % (age, schema["max_age_seconds"]))

    # VERDICT. INDETERMINATE is FORWARDED — visibly, as itself. Suppressing it would recreate the
    # dark-guard failure mode; promoting it to FAIL would page an operator for the wrong thing.
    if env["verdict"] == "INDETERMINATE":
        return Decision(FORWARD, "INDETERMINATE",
                        "the detector could not complete its measurement (run_outcome=%s)" % env["run_outcome"])
    if env["verdict"] == "PASS":
        return Decision(REFUSE, "PASS", "measured clean — nothing to page")
    return Decision(FORWARD, "FAIL", "measured and breaching (run_outcome=%s)" % env["run_outcome"])


def render_body(env, schema=None) -> str:
    """Render an alert body from `evidence` and NOTHING ELSE.

    D2 is structural, not stylistic: the old body asserted "SLO-ordered, so majors were served
    first; the shortfall is the long-tail overflow" — the inverse of what the code does, and it
    was never true. A body built here cannot assert a mechanism, because the only strings it can
    reach are evidence values, and the schema caps those at `max_prose_words`.
    """
    schema = schema if schema is not None else load_schema()
    lines = [
        "%s %s" % (env["verdict"], env["detector"]),
        "run=%s outcome=%s" % (env["run_id"], env["run_outcome"]),
        "window=%s..%s produced=%s" % (env["observation_window"]["from"],
                                       env["observation_window"]["to"], env["produced_at"]),
    ]
    for k in sorted(env["evidence"]):
        lines.append("%s=%s" % (k, env["evidence"][k]))
    return "\n".join(lines)


def _self_test() -> int:
    fails = 0

    def ck(label, got, want):
        nonlocal fails
        ok = got == want
        if not ok:
            fails += 1
        print("  %s %s — got %r, want %r" % ("PASS" if ok else "FAIL", label, got, want))

    try:
        schema = load_schema()
    except Exception as e:
        print("  could not read the schema: %s" % e)
        print("DETECTOR_ENVELOPE_VERDICT=INDETERMINATE")
        return 3

    print("detector_envelope --self-test")
    now = datetime(2026, 8, 22, 6, 53, 19, tzinfo=timezone.utc)

    def env(**over):
        base = {
            "schema_version": 1, "detector": "directional-label-capacity",
            "verdict": "FAIL", "run_id": "r1",
            "run_started_at": "2026-08-22T02:33:26Z", "run_outcome": "complete",
            "produced_at": "2026-08-22T03:20:01Z",
            "observation_window": {"from": "2026-08-22T02:33:26Z", "to": "2026-08-22T03:20:01Z"},
            "evidence": {"unreached_in_danger": "BINANCE,BITGET,BYBIT", "budget_min": 210},
        }
        base.update(over)
        return base

    ck("conforming envelope validates", validate(env(), schema), [])
    ck("a complete+FAIL run forwards", decide(env(), "r1", now, schema).action, FORWARD)
    ck("a stale run_id is REFUSED", decide(env(), "r2", now, schema).action, REFUSE)
    ck("a stale run_id refuses as INDETERMINATE, never FAIL",
       decide(env(), "r2", now, schema).verdict, "INDETERMINATE")
    ck("an unknown current run is REFUSED", decide(env(), None, now, schema).action, REFUSE)
    over_age = env(produced_at="2026-08-21T00:00:00Z")
    ck("an over-age envelope is REFUSED", decide(over_age, "r1", now, schema).action, REFUSE)
    ck("an over-age refusal names the bound",
       "past the" in decide(over_age, "r1", now, schema).reason, True)
    # The forcing rule, on the wire.
    stopped = env(run_outcome="stopped", verdict="FAIL")
    ck("a stopped run claiming FAIL is non-conforming", len(validate(stopped, schema)) > 0, True)
    ck("a stopped run's INDETERMINATE is conforming",
       validate(env(run_outcome="stopped", verdict="INDETERMINATE"), schema), [])
    ck("an INDETERMINATE envelope is FORWARDED, not silenced",
       decide(env(run_outcome="stopped", verdict="INDETERMINATE"), "r1", now, schema).action, FORWARD)
    ck("an INDETERMINATE envelope pages as INDETERMINATE, never FAIL",
       decide(env(run_outcome="stopped", verdict="INDETERMINATE"), "r1", now, schema).verdict, "INDETERMINATE")
    ck("a PASS envelope pages nobody", decide(env(verdict="PASS"), "r1", now, schema).action, REFUSE)
    # D2 — prose in evidence is rejected at the field that exists to prevent it.
    prose = env(evidence={"note": "SLO-ordered, so majors were served first; the shortfall is the long-tail overflow"})
    ck("mechanism prose in evidence is non-conforming", len(validate(prose, schema)) > 0, True)
    ck("the rendered body contains no mechanism prose",
       "long-tail overflow" in render_body(env(), schema), False)
    ck("the rendered body carries the measured values",
       "unreached_in_danger=BINANCE,BITGET,BYBIT" in render_body(env(), schema), True)
    ck("missing required field is caught", len(validate({"verdict": "PASS"}, schema)) > 0, True)
    # PINNED LITERALLY. Reading the expectation out of the schema would mean a schema that drops a
    # field silently stops being checked — measured: a mutation removing `run_id` left both this
    # self-test and the TS suite fully green.
    ck("the schema requires EXACTLY the contracted field set",
       sorted(schema["required_fields"]),
       sorted(["schema_version", "detector", "verdict", "run_id", "run_started_at",
               "run_outcome", "produced_at", "observation_window", "evidence"]))

    if fails:
        print("SELF-TEST: %d failing check(s)" % fails)
        print("DETECTOR_ENVELOPE_VERDICT=FAIL")
        return 1
    print("DETECTOR_ENVELOPE_VERDICT=PASS")
    return 0


if __name__ == "__main__":
    sys.exit(_self_test() if "--self-test" in sys.argv else _self_test())
